src/coreclr/jit/optimizer.cpp

```diff
@@ -7030,6 +7030,18 @@ bool Compiler::optIsProfitableToHoistTree(GenTree* tree, unsigned lnum)
         {
             availRegCount += CNT_CALLEE_TRASH_FLOAT - 1;
         }
+#if defined(TARGET_XARCH) && defined(FEATURE_SIMD)
+        else if ((varTypeIsSIMD(tree) && tree->TypeIs(TYP_SIMD32, TYP_SIMD64)) ||
+                 (tree->OperIsHWIntrinsic() && (tree->AsHWIntrinsic()->GetSimdSize() >= 32)))
+        {
+            // SIMD32/64 always spill due to the upper bits being volatile (callee trash)
+            // For now, avoid hoisting such constants when a call is present
+
+            // TODO-XArch-CQ: We should account for the spill cost as part of hoist determination
+            return false;
+        }
```

Comment:
> varTypeIsSIMD(tree) is redundant with tree->TypeIs(TYP_SIMD32, TYP_SIMD64).

`varTypeIsSIMD(tree)` is cheaper and ensures the common case is 1 check (just checking the `varTypeClassification`).

> Why have tree->OperIsHWIntrinsic() branch as well?

There are nodes that produce scalar results or small types but which still operate on V256/V512. It's an `||` condition so its handling either `node that produces a simd32/64` or `hwintrinsic node that uses simd32/64`

> This will not work for tree that contain SIMD expression but do not evaluate to a SIMD/HWI.

For some arbitrary block copies or other operations, potentially. Most usages of SIMD32/64 are done via hwintrinsics, so this is getting the important bits.

> The comment mentions ...such constants..., but the code does not contain checks for constants, so it is confusing.

Will fix.

Reply:
> For some arbitrary block copies or other operations, potentially

The concern is not about nodes that implicitly use SIMD, but cases like below:
```
ADD ; tree
  HWI_TREE_1
  HWI_TREE_2
```
One will notice that the rest of the method uses costs, which are calculated recursively.

This sort of conditioning will make hoisting sensitive to small changes to trees. E. g. you could have had that `ADD` hoisted before, but now something (e. g. the user) exposes `HWI_TREE`s, and they are not hoisted anymore.